drivers/rtc/rtc_xmc4xxx.c

```diff
@@ -0,0 +1,266 @@
+	unsigned int key;
+	int alarm = 0;
+	uint32_t event;
+
+	if (id != 0) {
+		return -EINVAL;
+	}
+
+	key = irq_lock();
+	event = XMC_SCU_INTERUPT_GetEventStatus();
+
+	if ((event & XMC_SCU_INTERRUPT_EVENT_RTC_ALARM) != 0) {
+		alarm = 1;
+		XMC_SCU_INTERRUPT_ClearEventStatus(XMC_SCU_INTERRUPT_EVENT_RTC_ALARM);
+	}
+	irq_unlock(key);
+
+	return alarm;
+}
+
+static int rtc_xmc4xxx_alarm_set_callback(const struct device *dev, uint16_t id,
+					  rtc_alarm_callback callback, void *user_data)
+{
+	struct rtc_xmc4xxx_data *dev_data = dev->data;
+	unsigned int key;
+
+	if (id != 0) {
+		return -EINVAL;
+	}
+
+	key = irq_lock();
+	dev_data->alarm_callback = callback;
+	dev_data->alarm_user_data = user_data;
+	irq_unlock(key);
+
+	if (dev_data->alarm_callback) {
+		XMC_SCU_INTERRUPT_EnableEvent(XMC_SCU_INTERRUPT_EVENT_RTC_ALARM);
+	} else {
+		XMC_SCU_INTERRUPT_DisableEvent(XMC_SCU_INTERRUPT_EVENT_RTC_ALARM);
+	}
+
+	return 0;
+}
+#endif /* CONFIG_RTC_ALARM */
+
+#ifdef CONFIG_RTC_UPDATE
+static int rtc_xmc4xxx_update_set_callback(const struct device *dev, rtc_update_callback callback,
+					   void *user_data)
+{
+	struct rtc_xmc4xxx_data *dev_data = dev->data;
+	unsigned int key;
+
+	key = irq_lock();
+	dev_data->update_callback = callback;
+	dev_data->update_user_data = user_data;
+	irq_unlock(key);
+
+	if (dev_data->update_callback) {
+		XMC_RTC_EnableEvent(XMC_RTC_EVENT_PERIODIC_SECONDS);
+		XMC_SCU_INTERRUPT_EnableEvent(XMC_SCU_INTERRUPT_EVENT_RTC_PERIODIC);
+	} else {
+		XMC_SCU_INTERRUPT_DisableEvent(XMC_SCU_INTERRUPT_EVENT_RTC_PERIODIC);
+		XMC_RTC_DisableEvent(XMC_RTC_EVENT_PERIODIC_SECONDS);
+	}
+
+	return 0;
+}
+#endif /* CONFIG_RTC_UPDATE */
+
+static const struct rtc_driver_api rtc_xmc4xxx_driver_api = {
+	.set_time = rtc_xmc4xxx_set_time,
+	.get_time = rtc_xmc4xxx_get_time,
+#ifdef CONFIG_RTC_ALARM
+	.alarm_get_supported_fields = rtc_xmc4xxx_alarm_get_supported_fields,
+	.alarm_set_time = rtc_xmc4xxx_alarm_set_time,
+	.alarm_get_time = rtc_xmc4xxx_alarm_get_time,
+	.alarm_is_pending = rtc_xmc4xxx_alarm_is_pending,
+	.alarm_set_callback = rtc_xmc4xxx_alarm_set_callback,
+#endif
+#ifdef CONFIG_RTC_UPDATE
+	.update_set_callback = rtc_xmc4xxx_update_set_callback,
+#endif
+};
+
+#if defined(CONFIG_RTC_ALARM) || defined(CONFIG_RTC_UPDATE)
+static void rtc_xmc4xxx_irq_config(void)
+{
+	/* RTC and watchdog share the same interrupt. Shared interrupts must */
+	/* be enabled if WDT is enabled and RTC is using alarm or update feature */
+	IRQ_CONNECT(DT_INST_IRQN(0), DT_INST_IRQ(0, priority), rtc_xmc4xxx_isr,
+		    DEVICE_DT_INST_GET(0), 0);
+	irq_enable(DT_INST_IRQN(0));
+}
+#endif
+
+static int rtc_xmc4xxx_init(const struct device *dev)
+{
+	const struct tm stdtime = { .tm_mday = 1, .tm_year = 1900};
+
+	if (!XMC_RTC_IsRunning()) {
+		if (!XMC_SCU_HIB_IsHibernateDomainEnabled()) {
+			XMC_SCU_HIB_EnableHibernateDomain();
+		}
+
+		XMC_RTC_SetPrescaler(RTC_XMC4XXX_DEFAULT_PRESCALER);
+		XMC_RTC_SetTimeStdFormat(&stdtime);
+
+		XMC_RTC_Start();
+	}
```

Comment:
The RTC should not be started and set to a random value at init, time is set manually using the rtc_set_time api :)